File: solutions/dart/matching-brackets/1/lib/matching_brackets.dart
class MatchingBrackets {
  bool isPaired(String str) {
    var stack = [];
    var pairs = {'}': '{', ')': '(', ']': '['};

    for (var c in str.runes) {
      var symbol = String.fromCharCode(c);
      if (pairs.containsValue(symbol)) {
        stack.add(symbol);
      } else if (pairs.containsKey(symbol)) {
        if (stack.isEmpty || stack.removeLast() != pairs[symbol]) {
          return false;
        }
      }
    }

    return stack.isEmpty;
  }
}

File: solutions/python/matching-brackets/1/matching_brackets.py
def is_paired(input_string):
    stack=[]
    pairs={']':'[', ')':'(', '}':'{'}

    for c in input_string:
        if c in pairs.values():
            stack.append(c)
        elif c in pairs.keys():
            if len(stack)==0 or stack.pop()!=pairs[c]:
                return False
    
    return len(stack)==0
